File: api/crs/crs_prodAx.m
function b = crs_prodAx(A, x, b, nthreads)
%crs_prodAx Compute b=A*x for a sparse matrix A in CRS format,
% assuming that number of columns of A is equal to size(x,1).
%
%      b = crs_prodAx(A, x [,b])
% Computes b=A*x in serial.
%
%      b = crs_prodAx(A, x, b, nthreads)
% Computes b=A*x Testing nthreads OpenMP threads, where nthreads is an int32.
% In this mode, there are implicit barriers in the function.
%
%      b = crs_prodAx(A, x, b, [])
% Computes b=A*x locally within each OpenMP thread, assuming the
% parallel region has already been initialized. The argument b has
% been preallocated. In this mode, there is no implicit barrier.
%
%      b = crs_prodAx(A, x, b, nthreads, comm)
% also performs communication within the MPI communicator.
% It assumes that A is partitioned columnwise, and hence A*x is
% a partial sum on each process. An allreduce is performed on b
% within the communicator.
%
% Note that if A is partitioned rowwise, then b must be duplicated
% on all processes, and prodAx should be called without the communicator.
%
%      b = crs_prodAx(A, x, b, nthreads, comm, pgmsg [, pgsz])
% specifies an additiona message that should be summed during
% allreduce and strored in b. This is to piggy-back on allreduce
% to help aggregate small messages to optimize MPI communication.
%
% The compiled version reqiures A and x to be double precision.
% The M file works for both single and double precision.

% See http://www.netlib.org/linalg/html_templates/node98.html

%#codegen -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),int32(0)}
%#codegen crs_prodAx_ser -args {crs_matrix, coder.typeof(0, [inf,inf])}
%#codegen crs_prodAx_ser1 -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}

coder.inline('never');

if nargin==2;
    b = coder.nullcopy(zeros(A.nrows,size(x,2)));
else
    if size(b,1)<A.nrows || size(b,2)<size(x,2)
        m2c_error('crs_prodAx:BufferTooSmal', 'Buffer space for output b is too small.');
    end
end
ismt = nargin>=4 && ompGetNumThreads>1;

if nargin>3 && ~isempty(nthreads)
    %% Declare parallel region
    if ~ompGetNested && ismt && nthreads(1)>1
        OMP_begin_master
        m2c_warn('crs_prodAx:NestedParallel', ...
            'You are trying to use nested parallel regions. Solution may be incorrect.');
        OMP_end_master
    end
    
    OMP_begin_parallel(int32(nthreads(1)));
    
    %% Compute b=A*x
    b = crs_prodAx_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x,1)), ...
        b, int32(size(b,1)), A.nrows, int32(size(x, 2)), ompGetNumThreads>1);
    
    %% End parallel region
    OMP_end_parallel;
else
    %% Compute b=A*x
    b = crs_prodAx_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x,1)), ...
        b, int32(size(b,1)), A.nrows, int32(size(x, 2)), ismt);
end

function b = crs_prodAx_kernel(row_ptr, col_ind, val, ...
    x, x_m, b, b_m, nrows, nrhs, ismt)

coder.inline('never');
if ismt
    [istart, iend] = OMP_local_chunk(nrows);
else
    istart = int32(1); iend = nrows;
end

xoffset = int32(0); boffset = int32(0);
for k=1:nrhs
    for i=istart:iend
        t = 0.0;
        for j=row_ptr(i):row_ptr(i+1)-1
            t = t + val(j)*x(xoffset+col_ind(j));
        end
        b(boffset+i) = t;
    end
    xoffset = xoffset + x_m; boffset = boffset + b_m;
end

function test %#ok<DEFNU>
%!test
%! if ~exist(['crs_prodAx.' mexext], 'file')
%!    m=100; n = 20;
%! else
%!    m=10000; n = 2000;
%! end
%! tic; sp = sprand(m,n,0.5); x=rand(size(sp,2),2);
%! [is,js,vs] = find(sp);
%! fprintf(1, '\n\tGenerated random matrix in %g seconds\n', toc);
%! tic; b0 = sp*x;
%! fprintf(1, '\tComputed reference solution in %g seconds\n', toc);
%! tic; A = crs_matrix(int32(is), int32(js), vs, int32(size(sp,1)), int32(size(sp,2)));
%! fprintf(1, '\tConverted into crs_matrix in %g seconds\n', toc);
%! fprintf(1, '\tTesting serial: ');
%! tic; b1 = crs_prodAx(A, x);
%! fprintf(1, 'Done in %g seconds\n ', toc);
%! assert(norm(b0-b1)<=1.e-12);

%! b2 = zeros(size(sp,1),2);
%! for nthreads=int32([1 2 4 8])
%!     if nthreads>ompGetMaxThreads; break; end
%!     fprintf(1, '\tTesting %d thread(s): ', nthreads);
%!     tic; b2 = crs_prodAx(A, x, b2, nthreads);
%!     fprintf(1, 'Done in %g seconds\n ', toc);
%!     assert(norm(b0-b2)<=1.e-12);
%! end

File: api/crs/codegen/lib/crs_prodAx/crs_prodAx.c
#include "crs_prodAx.h"
#include "crs_prodAx_types.h"
#include "m2c.h"
#include "omp.h"

static void b_crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                                const emxArray_int32_T *col_ind,
                                const emxArray_real_T *val,
                                const emxArray_real_T *x, int x_m,
                                emxArray_real_T *b, int b_m, int nrows,
                                int nrhs);

static void crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                              const emxArray_int32_T *col_ind,
                              const emxArray_real_T *val,
                              const emxArray_real_T *x, int x_m,
                              emxArray_real_T *b, int b_m, int nrows, int nrhs,
                              boolean_T varargin_1);

static void m2c_error(void);

static void m2c_warn(void);

static void b_crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                                const emxArray_int32_T *col_ind,
                                const emxArray_real_T *val,
                                const emxArray_real_T *x, int x_m,
                                emxArray_real_T *b, int b_m, int nrows,
                                int nrhs)
{
  double t;
  int b_i;
  int boffset;
  int i;
  int i1;
  int j;
  int k;
  int xoffset;
  xoffset = -1;
  boffset = 0;
  for (k = 0; k < nrhs; k++) {
    for (i = 0; i < nrows; i++) {
      t = 0.0;
      b_i = row_ptr->data[i];
      i1 = row_ptr->data[i + 1] - 1;
      for (j = b_i; j <= i1; j++) {
        t += val->data[j - 1] * x->data[xoffset + col_ind->data[j - 1]];
      }
      b->data[boffset + i] = t;
    }
    xoffset += x_m;
    boffset += b_m;
  }
}

static void crs_prodAx_kernel(const emxArray_int32_T *row_ptr,
                              const emxArray_int32_T *col_ind,
                              const emxArray_real_T *val,
                              const emxArray_real_T *x, int x_m,
                              emxArray_real_T *b, int b_m, int nrows, int nrhs,
                              boolean_T varargin_1)
{
  double t;
  int b_i;
  int b_remainder;
  int chunk;
  int i;
  int iend;
  int istart;
  int j;
  int k;
  int threadID;
  if (varargin_1) {
    iend = omp_get_num_threads();
    if (iend == 1) {
      istart = 0;
      iend = nrows;
    } else {
      threadID = omp_get_thread_num();
      chunk = nrows / iend;
      b_remainder = nrows - iend * chunk;
      if (b_remainder < threadID) {
        iend = b_remainder;
      } else {
        iend = threadID;
      }
      istart = threadID * chunk + iend;
      iend = (istart + chunk) + (threadID < b_remainder);
    }
  } else {
    istart = 0;
    iend = nrows;
  }
  b_remainder = -1;
  threadID = -1;
  for (k = 0; k < nrhs; k++) {
    for (i = istart + 1; i <= iend; i++) {
      t = 0.0;
      chunk = row_ptr->data[i - 1];
      b_i = row_ptr->data[i] - 1;
      for (j = chunk; j <= b_i; j++) {
        t += val->data[j - 1] * x->data[b_remainder + col_ind->data[j - 1]];
      }
      b->data[threadID + i] = t;
    }
    b_remainder += x_m;
    threadID += b_m;
  }
}

static void m2c_error(void)
{
  M2C_error("crs_prodAx:BufferTooSmal",
            "Buffer space for output b is too small.");
}

static void m2c_warn(void)
{
  M2C_warn("crs_prodAx:NestedParallel", "You are trying to use nested parallel "
                                        "regions. Solution may be incorrect.");
}

void crs_prodAx(const CRS_Matrix *A, const emxArray_real_T *x,
                emxArray_real_T *b, int nthreads)
{
  int n;
  if ((b->size[0] < A->nrows) || (b->size[1] < x->size[1])) {
    m2c_error();
  }
  n = omp_get_nested();
  if (n == 0) {
    n = omp_get_num_threads();
    if ((n > 1) && (nthreads > 1)) {
#pragma omp master
      {
        m2c_warn();
      }

    }
  }
#pragma omp parallel default(shared) num_threads(nthreads)
  {
    n = omp_get_num_threads();
    crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b,
                      b->size[0], A->nrows, x->size[1], n > 1);
  }

}

void crs_prodAx_initialize(void)
{
}

void crs_prodAx_ser(const CRS_Matrix *A, const emxArray_real_T *x,
                    emxArray_real_T *b)
{
  int i;
  i = b->size[0] * b->size[1];
  b->size[0] = A->nrows;
  b->size[1] = x->size[1];
  emxEnsureCapacity_real_T(b, i);
  b_crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b,
                      A->nrows, A->nrows, x->size[1]);
}

void crs_prodAx_ser1(const CRS_Matrix *A, const emxArray_real_T *x,
                     emxArray_real_T *b)
{
  if ((b->size[0] < A->nrows) || (b->size[1] < x->size[1])) {
    m2c_error();
  }
  b_crs_prodAx_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b,
                      b->size[0], A->nrows, x->size[1]);
}

void crs_prodAx_terminate(void)
{
}
